### DISCUSSION TITLE
Texture difference between Chrome and Safari

### DISCUSSION THREAD

**Discussion by garciat** [gpuweb/gpuweb#5035] [Q&A]
I wrote this code based on other stuff I found online: https://garciat.com/webgpu-playground/posts/hdr/

I can't figure out why the texture rending on Safari always shows the grass texture on the cube.

I even tried deleting the grass texture altogether. Safari still renders it.

JS: https://github.com/Garciat/webgpu-playground/blob/main/src/posts/hdr/main.ts
WGSL: https://github.com/Garciat/webgpu-playground/blob/main/src/posts/hdr/shaders.wgsl

I am still learning graphics programming, so I wouldn't be surprised if there's a silly bug in the code.

Chrome (v131.0.6778.205)
<img width="400" alt="image" src="https://github.com/user-attachments/assets/f73c987a-7c66-498a-a224-b0fd18835110" />

Safari (v18.2)
<img width="400" alt="image" src="https://github.com/user-attachments/assets/cfe50dc7-17a9-46e1-b375-2d4eab1a9c0b" />

**Comment by garciat** [comment]
If I restart Safari, then the cube texture doesn't render at all.

I also tried my cat texture on the [textured cube sample](https://webgpu.github.io/webgpu-samples/sample/texturedCube/) without any code changes. I get the same issue.

So now I think that Safari fails on `copyExternalImageToTexture`.

I ruled out `createImageBitmap` being a problem by loading the cat image as a bitmap and rendering into a 2D canvas context. It worked out just fine.

The cat image I'm using contain HDR information; maybe Safari doesn't like its format.

**Reply by greggman** [reply]
Safari's WebGPU implementation is still unfinished (as of 2024-12-31) which is why it still requires you to enable it in the developer menu so it's certainly possible that somethings don't work yet.

The tests for `copyExternalImageToTexture` are here

[https://gpuweb.github.io/cts/standalone/?runnow=1&q=webgpu:web_platform,copyToTexture,*](https://gpuweb.github.io/cts/standalone/?runnow=1&q=webgpu:web_platform,copyToTexture,*)

running them locally it looks like Safari has issues with `premultipliedAlpha: true` for some formats. Of course there may be other issues.
